src/ray/util/logging.cc

```diff
@@ -0,0 +1,147 @@
+#include <cstdlib>
+#include <iostream>
+
+#include "ray/util/logging.h"
+
+#ifdef RAY_USE_GLOG
+#include "glog/logging.h"
+#endif
+
+namespace ray {
+
+// This is the default implementation of ray log,
+// which is independent of any libs.
+class CerrLog {
+ public:
+  CerrLog(int severity) : severity_(severity), has_logged_(false) {}
+
+  virtual ~CerrLog() {
+    if (has_logged_) {
+      std::cerr << std::endl;
+    }
+    if (severity_ == RAY_FATAL) {
+      PrintBackTrace();
+      std::abort();
+    }
+  }
+
+  std::ostream &Stream() {
+    has_logged_ = true;
+    return std::cerr;
+  }
+
+  template <class T>
+  CerrLog &operator<<(const T &t) {
+    if (severity_ != RAY_DEBUG) {
+      has_logged_ = true;
+      std::cerr << t;
+    }
+    return *this;
+  }
+
+ protected:
+  const int severity_;
+  bool has_logged_;
+
+  void PrintBackTrace() {
+#if defined(_EXECINFO_H) || !defined(_WIN32)
+    void *buffer[255];
+    const int calls = backtrace(buffer, sizeof(buffer) / sizeof(void *));
+    backtrace_symbols_fd(buffer, calls, 1);
+#endif
+  }
+};
+
+int RayLog::severity_threshold_ = RAY_INFO;
+
+#ifdef RAY_USE_GLOG
+using namespace google;
+
+// Glog's severity map.
+static int GetMappedSeverity(int severity) {
+  static int severity_map[] = {
+      // glog has no DEBUG level. It has verbose level but hard to adapt here.
+      GLOG_INFO,     // RAY_DEBUG
+      GLOG_INFO,     // RAY_INFO
+      GLOG_WARNING,  // RAY_WARNING
+      GLOG_ERROR,    // RAY_ERROR
+      GLOG_FATAL     // RAY_FATAL
+  };
+  // Ray log level starts from -1 (RAY_DEBUG);
+  return severity_map[severity + 1];
```

Comment:
up to you, but you could consider replacing the whole static int map with something like this might be cleaner:
```
switch(severity) {
case RAY_DEBUG:
  return GLOG_INFO;
case RAY_INFO:
  return GLOG_INFO;
case RAY_WARNING:
  return GLOG_WARNING;
// etc, and can have a fatal check in the default case
}
```

Reply:
I test the perf impact. Testing RAY_ERROR log for 100000 rounds takes from 2744 ms to 2763 ms, which is not too big. I will choose the switch statement.